### PR TITLE
Device sensor support (Accounting attributes)

### PR DESCRIPTION
**Pull request by fdurand**
# Description
Device sensor support in PacketFence (https://www.cisco.com/c/en/us/td/docs/switches/lan/catalyst3750/software/release/15-0_1_se/device_sensor/guide/sensor_guide.html)
Update device information based on radius accounting attributes.

# Impacts
-RADIUS Accounting

# Delete branch after merge
YES

# NEWS file entries
## New Features
* Support for Cisco Device sensor feature (Update node information based on accounting, like dhcp, lldp, cdp, http)

### CODE REVIEW

**Comment by fdurand** [comment]
Missing documentation

File: lib/pf/Switch/Cisco.pm [hunk 1]
```diff
@@ -1787,6 +1791,64 @@ sub cmnMacChangedNotificationTrapNormalizer {
...
+
+=cut
+
+sub acctVoipDetect {
+    my ( $self, $decoded) = @_;
+    $decoded->{'reason'} = 'node_modify';
+
+    if ($decoded->{'isPhone'}) {
+        my $node_attributes = node_attributes($decoded->{mac});
+        if ( $node_attributes->{'voip'} eq 'no') {
```
Comment (julsemaan) [inline lib/pf/Switch/Cisco.pm]: constant for no

File: html/pfappserver/lib/pfappserver/Form/Config/Switch.pm [hunk 2]
```diff
@@ -159,6 +159,25 @@ has_field 'VoIPDHCPDetect' =>
              help => 'Detect VoIP with the DHCP Fingerprint'},
   );
 
+has_field 'VoIPAccountingDetect' =>
+  (
+   type => 'Toggle',
+   label => 'VoIPAccountingDetect',
+   default => undef,
+   tags => { after_element => \&help,
+             help => 'Detect VoIP with the radius attributes contain in accounting'},
```
Comment (julsemaan) [inline html/pfappserver/lib/pfappserver/Form/Config/Switch.pm]: contained

File: html/pfappserver/lib/pfappserver/Form/Config/Switch.pm [hunk 3]
```diff
@@ -159,6 +159,25 @@ has_field 'VoIPDHCPDetect' =>
...
+  );
+
+has_field 'RadiusFingerprint' =>
+  (
+   type => 'Toggle',
+   label => 'RadiusFingerprint',
+   default => undef,
+   tags => { after_element => \&help,
+             help => 'Enable fingerprinting with radius attributes contain in accounting'},
```
Comment (julsemaan) [inline html/pfappserver/lib/pfappserver/Form/Config/Switch.pm]: contained

File: lib/pf/Switch/Cisco.pm [hunk 4]
```diff
@@ -1787,6 +1791,64 @@ sub cmnMacChangedNotificationTrapNormalizer {
...
+            require pf::role;
+            my $role_obj = new pf::role::custom();
+            my %autoreg_node_defaults = $role_obj->getNodeInfoForAutoReg($decoded);
+            $node_attributes = merge($node_attributes, \%autoreg_node_defaults);
+            my $apiclient = pf::client::getClient;
+            $apiclient->notify('modify_node', %{$node_attributes} );
+            $apiclient->notify('reevaluate_access', %{$decoded} ) if ( $node_attributes->{'voip'} eq 'no');
```
Comment (julsemaan) [inline lib/pf/Switch/Cisco.pm]: constant for no

Comment (julsemaan) [reply lib/pf/Switch/Cisco.pm]: Shouldn't we also set voip to yes after this ?

Comment (fdurand) [reply lib/pf/Switch/Cisco.pm]: Yes, the cisco_device_sensor will return isPhone to yes if it detect cdp/lldp voip flag.

**Comment by extrafu** [comment]
Please see with @julsemaan for Fingerbank Collector integration and ensure VoIP attributes are well recognized.

**Comment by fdurand** [comment]
Can i close it, since the collector do the job ?

**Comment by extrafu** [comment]
@fdurand I think we should keep the documentation which is great. I'll see where I can put it in the official one.

**Comment by extrafu** [comment]
Took the doc.

**Pull request closed without merging**